Sources/GoogleCloudProvider/Storage/StorageRequest.swift

```diff
@@ -18,48 +18,68 @@ extension HTTPHeaders {
+    let dateFormatter: DateFormatter
+
     var currentToken: OAuthAccessToken?
     var tokenCreatedTime: Date?
     
     init(httpClient: Client, oauth: OAuthRefreshable, project: String) {
         self.refreshableToken = oauth
         self.httpClient = httpClient
         self.project = project
+        self.responseDecoder = JSONDecoder()
+        self.dateFormatter = DateFormatter()
+
+        self.dateFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSZ"
+        self.responseDecoder.dateDecodingStrategy = .formatted(self.dateFormatter)
     }
     
     func send<GCM: GoogleCloudModel>(method: HTTPMethod, headers: HTTPHeaders = [:], path: String, query: String, body: HTTPBody = HTTPBody()) throws -> Future<GCM> {
-        // if oauth token is not expired continue as normal
+        return try withToken({ token in
+            return try self.send(method: method, headers: headers, path: path, query: query, body: body, accessToken: token.accessToken).flatMap({ response in
+                return try self.responseDecoder.decode(GCM.self, from: response.http, maxSize: 65_536, on: response)
+            })
+        })
+    }
 
-        if let token = currentToken, let created = tokenCreatedTime, refreshableToken.isFresh(token: token, created: created) {
-            return try _send(method: method, headers: headers, path: path, query: query, body: body, accessToken: token.accessToken)
-        } else {
-            return try refreshableToken.refresh().flatMap({ (newToken) in
+    func send(method: HTTPMethod, headers: HTTPHeaders = [:], path: String, query: String, body: HTTPBody = HTTPBody()) throws -> Future<Data> {
+        return try withToken({ token in
+            return try self.send(method: method, headers: headers, path: path, query: query, body: body, accessToken: token.accessToken).flatMap({ response in
+                return response.http.body.consumeData(on: response)
+            })
+        })
+    }
+
+    private func withToken<F>(_ closure: @escaping (OAuthAccessToken) throws -> Future<F>) throws -> Future<F>{
+        guard let token = currentToken, let created = tokenCreatedTime, refreshableToken.isFresh(token: token, created: created) else {
+            return try refreshableToken.refresh().flatMap({ newToken in
                 self.currentToken = newToken
                 self.tokenCreatedTime = Date()
-                return try self._send(method: method, headers: headers, path: path, query: query, body: body, accessToken: newToken.accessToken)
+
+                return try closure(newToken)
             })
         }
+
+        return try closure(token)
     }
-    
-    private func _send<GCM: GoogleCloudModel>(method: HTTPMethod, headers: HTTPHeaders, path: String, query: String, body: HTTPBody, accessToken: String) throws -> Future<GCM> {
+
+    private func send(method: HTTPMethod, headers: HTTPHeaders, path: String, query: String, body: HTTPBody, accessToken: String) throws -> Future<Response> {
```

Comment:
Can we mark this send as `_send` it was hard to actually read in the public send functions which send it's actually calling since there's now 3 send methods that slightly differ.